sdk/src/Core/Internal/Utils/HostEndPoint.cs

```diff
@@ -0,0 +1,102 @@
+﻿using System;
+using System.Linq;
+using System.Net;
+using System.Net.Sockets;
+using Amazon.Runtime.Internal.Util;
+
+namespace Amazon.XRay.Recorder.Core.Internal.Utils
+{
+	public class HostEndPoint
+	{
+		private const int CACHE_TTL = 60;	//Seconds to consider a cached dns response valid
+		private IPEndPoint _ipCache = null;
+		private DateTime? _timestampOfLastIPCacheUpdate = null;
+		private static readonly Logger _logger = Logger.GetLogger(typeof(HostEndPoint));
+		private HostEndPoint(string host, int port)
+		{
+			Host = host;
+			Port = port;
+		}
+		public string Host { get; private set; }
+		public int Port { get; private set; }
+
+
+		private bool isIPCacheValid()
+		{
+			if (_ipCache == null)
+			{
+				return false;
+			}
+
+			if (_timestampOfLastIPCacheUpdate is DateTime lastTimestamp)
+			{
+				return DateTime.Now.Subtract(lastTimestamp).TotalSeconds < CACHE_TTL;
+			}
+			else
+			{
+				return false;
+			}
+		}
+
+		public IPEndPoint GetIPEndPoint()
```

Comment:
7572b7b4d333bfb6fcc408c01fdc7fa4bcdb50c3 addresses two threads concurrently updating the cache. That is the only write operation on the cache.

If two threads tries to update the cache concurrently only one will do so. The other will immediately return, opening up for the possibility that it will return a dirty cache. This is however not a problem due to we can't ever really guarantee the cache to not be dirty because the cached IP can become dirty before the TTL run out and will allow the critical path will still be speedy and not waiting on the DNS resolution of the first thread to complete.